### PR TITLE
[Merged by Bors] - feat(Topology/Order/ScottTopology): Characterise the Scott Topology of a Complete Linear Order

### PR DESCRIPTION
**Pull request by mans0954**
This PR gives a convenient characterisation of the Scott Topology for a Complete Linear Order

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 1]
```diff
@@ -236,6 +236,9 @@ lemma isUpperSet_of_isOpen : IsOpen s → IsUpperSet s := fun h ↦
 lemma isLowerSet_of_isClosed : IsClosed s → IsLowerSet s := fun h ↦
   (isClosed_iff_isLowerSet_and_dirSupClosed.mp h).left
 
+lemma DirSupClosed_of_isClosed : IsClosed s → DirSupClosed s := fun h ↦
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:239]:
```suggestion
lemma dirSupClosed_of_isClosed : IsClosed s → DirSupClosed s := fun h ↦
```

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 2]
```diff
@@ -291,6 +294,34 @@ instance (priority := 90) : T0Space α :=
...
+  · intro hU
+    rcases eq_empty_or_nonempty Uᶜ with eUc | neUc
+    · exact Or.inr (compl_empty_iff.mp eUc)
+    · apply Or.inl
+      let b := sSup Uᶜ
+      use b
+      rw [eq_compl_comm, le_antisymm_iff]
+      exact ⟨IsLowerSet.Iic_subset (isLowerSet_of_isClosed (isClosed_compl_iff.mpr hU))
+        ((dirSupClosed_iff_forall_sSup.mp (DirSupClosed_of_isClosed (isClosed_compl_iff.mpr hU)))
+        neUc (IsChain.directedOn (isChain_of_trichotomous Uᶜ)) (le_refl Uᶜ)),
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:314]:
```suggestion
        neUc (isChain_of_trichotomous Uᶜ).directedOn le_rfl,
```

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 3]
```diff
@@ -291,6 +294,34 @@ instance (priority := 90) : T0Space α :=
...
+  constructor
+  · intro hU
+    rcases eq_empty_or_nonempty Uᶜ with eUc | neUc
+    · exact Or.inr (compl_empty_iff.mp eUc)
+    · apply Or.inl
+      let b := sSup Uᶜ
+      use b
+      rw [eq_compl_comm, le_antisymm_iff]
+      exact ⟨IsLowerSet.Iic_subset (isLowerSet_of_isClosed (isClosed_compl_iff.mpr hU))
+        (dirSupClosed_iff_forall_sSup.mp (dirSupClosed_of_isClosed (isClosed_compl_iff.mpr hU))
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:313]:
```suggestion
        (dirSupClosed_iff_forall_sSup.mp (dirSupClosed_of_isClosed  hU.isClosed_compl)
```

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 4]
```diff
@@ -291,6 +294,34 @@ instance (priority := 90) : T0Space α :=
...
+        neUc (isChain_of_trichotomous Uᶜ).directedOn le_rfl),
+        fun  _ ha ↦ CompleteLattice.le_sSup Uᶜ _ ha⟩
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:315]:
```suggestion
        fun  _ ha ↦ le_sSup Uᶜ _ ha⟩
```

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 5]
```diff
@@ -291,6 +294,34 @@ instance (priority := 90) : T0Space α :=
...
+  · intro H
+    rcases H with ⟨a,ha⟩ | hU
+    · rw [← isClosed_compl_iff, ha, compl_compl]
+      exact isClosed_Iic
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:319]:
```suggestion
    · exact isClosed_Iic.isOpen_compl
```

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 6]
```diff
@@ -291,6 +294,34 @@ instance (priority := 90) : T0Space α :=
...
+    · apply Or.inl
+      let b := sSup Uᶜ
+      use b
+      rw [eq_compl_comm, le_antisymm_iff]
+      exact ⟨IsLowerSet.Iic_subset (isLowerSet_of_isClosed (isClosed_compl_iff.mpr hU))
+        (dirSupClosed_iff_forall_sSup.mp (dirSupClosed_of_isClosed (isClosed_compl_iff.mpr hU))
+        neUc (isChain_of_trichotomous Uᶜ).directedOn le_rfl),
+        fun  _ ha ↦ CompleteLattice.le_sSup Uᶜ _ ha⟩
+  · intro H
+    rcases H with ⟨a,ha⟩ | hU
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/ScottTopology.lean:317]:
```suggestion
  · rintro ⟨a,ha⟩ | rfl
```

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 7]
```diff
@@ -291,6 +294,32 @@ instance (priority := 90) : T0Space α :=
...
+
+lemma isOpen_iff_Iic_compl_or_univ (U : Set α) :
+    IsOpen U ↔ (∃ (a : α), U = (Iic a)ᶜ) ∨ U = univ := by
+  constructor
+  · intro hU
+    rcases eq_empty_or_nonempty Uᶜ with eUc | neUc
+    · exact Or.inr (compl_empty_iff.mp eUc)
+    · apply Or.inl
+      let b := sSup Uᶜ
+      use b
```
Comment (riccardobrasca) [inline Mathlib/Topology/Order/ScottTopology.lean:310]:
```suggestion
      use sSup Uᶜ
```
It seems you use `b` only once.

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

File: Mathlib/Topology/Order/ScottTopology.lean [hunk 8]
```diff
@@ -313,8 +313,7 @@ lemma isOpen_iff_Iic_compl_or_univ (U : Set α) :
         (dirSupClosed_iff_forall_sSup.mp (dirSupClosed_of_isClosed  hU.isClosed_compl)
         neUc (isChain_of_trichotomous Uᶜ).directedOn le_rfl),
         fun  _ ha ↦ le_sSup Uᶜ _ ha⟩
-  · intro H
-    rcases H with ⟨a,ha⟩ | hU
+  · rintro ⟨a,ha⟩ | rfl
```
Comment (mans0954) [inline Mathlib/Topology/Order/ScottTopology.lean]: @YaelDillies the closest I can get is:
```
  · intro H
    rcases H with ⟨a,rfl⟩ | rfl
    · exact isClosed_Iic.isOpen_compl
    · exact isOpen_univ
```

Comment (mans0954) [reply Mathlib/Topology/Order/ScottTopology.lean]: This works:
```
  · rintro (⟨a,rfl⟩ | rfl)
    · exact isClosed_Iic.isOpen_compl
    · exact isOpen_univ
```

**Pull request closed without merging**